### PR TITLE
Use ChaCha only if prioritized by clnt

### PR DESCRIPTION
**Pull request by tmshort**
IFF the client has ChaCha first, and server cipher priority is used,
and the new SSL_OP_PRIORITIZE_CHACHA_FOR_MOBILE option is used,
then reprioritize ChaCha above everything else. This way, A matching
ChaCha cipher will be selected if there is a match. If no ChaCha ciphers
match, then the other ciphers are used.

DO NOT SQUASH THE TWO COMMITS; ONE CAN BE BACK-PORTED TO 
1.1.0, THE OTHER CANNOT.

##### Checklist

- [X] documentation is added or updated
- [X] tests are added or updated

### CODE REVIEW

**Comment by richsalz** [comment]
We recently heard from several companies that performance is a concern, and this patch helps address that.  Google had a patch that added "equivalence classes" to the ciphers string, but it was kind of incomplete.  And the cipher string syntax is already cluttered. So while adding this as an option, only in the code isn't as comprehensive, it's cleaner and can be supported in the conf file.

**Comment by davidben** [comment]
@richsalz I'm curious what you think is incomplete about the equipreference groups.

This design potentially runs into issues if new ciphers are added in front. I believe the list does skip unknown ciphers, so that clears the immediate extensibility failure, but one could imagine, e.g., SCSV ciphers being in front. It also breaks if, for whatever other reason, none of the ChaCha ciphers are suitable. Maybe all common ChaCha ciphers are for the wrong version. Or maybe they're all ECDHE and you have no curves in common.

**Comment by davidben** [comment]
(Oh, sorry, scratch the second comment. I'd missed that you do pull in the rest of the ciphers. Though looking at the first cipher is still a little fragile.)

**Comment by richsalz** [comment]
As I recall, you document that the equiv class means you can't sort after it?  I forget.

**Comment by davidben** [comment]
Ah, that's true. Adding an equivalence class disables most opcodes. But it's pretty clear how to make them work. I've been meaning to do it but the need hadn't really come up yet.

**Comment by levitte** [comment]
Small nit to pick: please drop `_FOR_MOBILE` / `ForMobile`...  the library has no way of knowing anyway.

**Comment by tmshort** [comment]
> (Oh, sorry, scratch the second comment. I'd missed that you do pull in the rest of the ciphers. Though looking at the first cipher is still a little fragile.)

This just reorders the server ciphers. I've seen other patches for similar functionality that discard all other ciphers, and as you point out that won't work. Most (all?) mobile clients indicate their preference for ChaCha by putting it at the beginning of their cipher list, so that's the key we look at.

SCSV is only present in client hellos; it won't matter how the server ciphers are ordered. I am not aware of any client that puts SCSV in the front.

**Comment by tmshort** [comment]
> Small nit to pick: please drop _FOR_MOBILE / ForMobile... the library has no way of knowing anyway.

It was to indicate intended use of the option; it was a last minute addition.

**Comment by tmshort** [comment]
The last commit fixes the nits found thus far.

File: ssl/s3_lib.c [hunk 1]
```diff
@@ -4148,6 +4151,34 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
+            c = sk_SSL_CIPHER_value(clnt, 0);
+            if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                /* Is client preferred, reprioritize... */
+                new_chacha = sk_SSL_CIPHER_new_null();
+                if (new_chacha != NULL) {
+                    /* Put ChaCha at the top */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
```
Comment (InfoHunter) [inline ssl/s3_lib.c]: This `if` 's `{ }` seems not necessary...

File: ssl/s3_lib.c [hunk 2]
```diff
@@ -4148,6 +4151,34 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                        }
+                    }
+                    /* Pull in the rest */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc != SSL_CHACHA20POLY1305) {
```
Comment (InfoHunter) [inline ssl/s3_lib.c]: as above

File: ssl/s3_lib.c (hunk 1 above)
Comment (tmshort) [reply ssl/s3_lib.c]: Yes, the coding standard indicates that braces are not needed for one-statement blocks.

Comment (richsalz) [reply ssl/s3_lib.c]: HOWEVER, if the code is more clear, feel free to put them. :)

Comment (tmshort) [reply ssl/s3_lib.c]: That thought did cross my mind...

**Comment by InfoHunter** [comment]
If this could be merged into 1.1.1, then many people don't need to struggle with maintaining a patch...

File: ssl/s3_lib.c (hunk 1 above)
Comment (tmshort) [reply ssl/s3_lib.c]: I'll hold off doing anything here, unless there are other required changes.

File: doc/man3/SSL_CTX_set_options.pod [hunk 3]
```diff
@@ -306,9 +314,11 @@ L<dhparam(1)>
 The attempt to always try to use secure renegotiation was added in
 Openssl 0.9.8m.
 
+B<SSL_OP_PRIORITIZE_CHACHA> was assed in OpenSSL 1.1.1.
```
Comment (dot-asm) [inline doc/man3/SSL_CTX_set_options.pod]: This doesn't comply with Oxford dictionary, which offers to meanings, to call and to act as.

Comment (dot-asm) [reply doc/man3/SSL_CTX_set_options.pod]: I mean two meanings for the verb.

File: include/openssl/ssl.h [hunk 4]
```diff
@@ -338,6 +338,9 @@ typedef int (*SSL_verify_cb)(int preverify_ok, X509_STORE_CTX *x509_ctx);
 # define SSL_OP_ALLOW_UNSAFE_LEGACY_RENEGOTIATION        0x00040000U
 /* Disable encrypt-then-mac */
 # define SSL_OP_NO_ENCRYPT_THEN_MAC                      0x00080000U
+/* Prioritize Chacha20Poly1305 when client does.
+ * Modifies SSL_OP_CIPHER_SERVER_PREFERENCE */
+# define SSL_OP_PRIORITIZE_CHACHA                        0x00200000U
```
Comment (dot-asm) [inline include/openssl/ssl.h]: What's rationale behind choice of constant? Basically there is contradiction. If rationale is that 0x00100000 was removed in 1.1.0, while 0x00200000 in 1.0.2, then it doesn't add up with previous constant. I mean if this is the reason that one probably should adjust even previous constant... But formally speaking removed constant should not be a problem, because you are forced to recompile when upgrading to 1.1.x and it's sufficient to just redefine corresponding macros as zeros. I mean at source code level they won't cause backward compatibility clash...

File: ssl/s3_lib.c [hunk 5]
```diff
@@ -4148,6 +4151,34 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+                new_chacha = sk_SSL_CIPHER_new_null();
+                if (new_chacha != NULL) {
+                    /* Put ChaCha at the top */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                        }
+                    }
+                    /* Pull in the rest */
```
Comment (dot-asm) [inline ssl/s3_lib.c]: Why does one have to compose the rest of the list?

File: doc/man3/SSL_CTX_set_options.pod (hunk 3 above)
Comment (tmshort) [reply doc/man3/SSL_CTX_set_options.pod]: In my own defense, 's' and 'd' are next to each other on the keyboard, and I did type a real word. :)

File: include/openssl/ssl.h (hunk 4 above)
Comment (tmshort) [reply include/openssl/ssl.h]: The rational is that I have another outstanding PR #3910 using 0x00100000

File: ssl/s3_lib.c (hunk 5 above)
Comment (tmshort) [reply ssl/s3_lib.c]: Although both lists may have ChaCha20, they may not be the same cipher suite; so the rest of the cipher list is necessary. This prevents a failure case (no matching cipher) that would have otherwise succeeded.

File: include/openssl/ssl.h (hunk 4 above)
Comment (tmshort) [reply include/openssl/ssl.h]: (There was a prior issue with values that were in SSL_OP_ALL in 1.1.0 and unused, but those values have been marked reserved until 1.2.0)

File: ssl/s3_lib.c (hunk 5 above)
Comment (tmshort) [reply ssl/s3_lib.c]: (I have seen other ChaCha20 priority patches that have this failure mode because they don't consider the non-ChaCha ciphers their processing)

Comment (dot-asm) [reply ssl/s3_lib.c]: Git it.

Comment (dot-asm) [reply ssl/s3_lib.c]: :-) I missed one centimeter too! I mean I meant to write "got it"...

File: ssl/s3_lib.c [hunk 6]
```diff
@@ -4148,6 +4151,32 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     if (s->options & SSL_OP_CIPHER_SERVER_PREFERENCE || tls1_suiteb(s)) {
         prio = srvr;
         allow = clnt;
+#ifndef OPENSSL_NO_CHACHA
+        /* If ChaCha is at the top of the client preference list,
+           temporarily prioritize ChaCha ciphers in the servers list. */
+        if (s->options & SSL_OP_PRIORITIZE_CHACHA && sk_SSL_CIPHER_num(clnt) > 0) {
```
Comment (dot-asm) [inline ssl/s3_lib.c]: One can wonder if it's appropriate to do so in tls1_suiteb case... Well, it might happen that Chacha won't make into server list in such case, but then making a copy of cipher list would be waste of resources...

File: ssl/s3_lib.c [hunk 7]
```diff
@@ -4148,6 +4151,32 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+            c = sk_SSL_CIPHER_value(clnt, 0);
+            if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                /* Is client preferred, reprioritize... */
+                new_chacha = sk_SSL_CIPHER_new_null();
```
Comment (dot-asm) [inline ssl/s3_lib.c]: There is now sk_\*_reserve that should be preferred to use here. [And it would make sense to check for its return value.] Though if goal is to back-port it, then of course one can't use it. But one can and probably should add it afterward. Or as separate commit, so that one could be cherry-picked, but not the other...

File: ssl/s3_lib.c [hunk 8]
```diff
@@ -4148,6 +4151,32 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc == SSL_CHACHA20POLY1305)
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                    }
+                    /* Pull in the rest */
+                    for (i = 0; i < sk_SSL_CIPHER_num(srvr); i++) {
+                        c = sk_SSL_CIPHER_value(srvr, i);
+                        if (c->algorithm_enc != SSL_CHACHA20POLY1305)
+                            sk_SSL_CIPHER_push(new_chacha, c);
+                    }
```
Comment (dot-asm) [inline ssl/s3_lib.c]: One can actually do it in single loop: `if (c->algorithm_enc == SSL_CHACHA20POLY1305) sl_SSL_CIPHER_unsift(new_chacha, c); else sk_SSL_CIPHER_push(new_chacha, c);`. I also wonder if compiler keeps calling sk_SSL_CIPHER_num for each iteration. I mean maybe it makes sense to ensure that it's called only once...

File: ssl/s3_lib.c (hunk 7 above)
Comment (dot-asm) [inline ssl/s3_lib.c]: There is now sk_\*_reserve that should be preferred to use here. [And it would make sense to check for its return value.] Though if goal is to back-port it, then of course one can't use it. But one can and probably should add it afterward. Or as separate commit, so that one could be cherry-picked, but not the other...

Comment (dot-asm) [reply ssl/s3_lib.c]: ??? Can browser play trick on you and submit same message twice with more than ~10 minutes interval? I mean this is duplicate...

File: ssl/s3_lib.c (hunk 8 above)
Comment (dot-asm) [reply ssl/s3_lib.c]: Yes, it calls sk_SSL_CIPHER_num for each iteration.

Comment (tmshort) [reply ssl/s3_lib.c]: So, how would one prevent it from being called?

File: ssl/s3_lib.c (hunk 6 above)
Comment (tmshort) [reply ssl/s3_lib.c]: One would have to iterate through the stack to see if there's ChaCha to make the decision to re-order. But is that worth it?

Comment (dot-asm) [reply ssl/s3_lib.c]: Here is my line of reasoning. Note that tls1_suiteb effectively takes precedence over SSL_OP_CIPHER_SERVER_PREFERENCE(*). It's natural to expect that it should take precedence even over SSL_OP_PRIORITIZE_CHACHA. So that for this reason alone one can wonder if check for SSL_OP_PRIORITIZE_CHACHA should be performed only in non-suiteb case. But on the other hand chances are that in suiteb case Chacha isn't even there(**) so one can use tls1_suiteb as criteria for not executing effectively no-op code.

(*) In sense that tls1_suiteb prioritizes server ciphers even without SSL_OP_CIPHER_SERVER_PREFERENCE. In the context one can even wonder if flipping the two in if condition would make code more readable. Or even `if (tls1_suiteb) {} else if (SSL_OP_CIPHER_SERVER_PREFERENCE) {} else {}`.

(**) This seems to be untrue in sense that if I s_server -suiteB_128 I can still connect with s_client -cipher ECDHE-RSA-CHACHA20-POLY1305. But nevertheless, first reason is probably sufficient by itself.

Comment (dot-asm) [reply ssl/s3_lib.c]: Well, there is also chance that I'm missing something specific about tls1_suiteb. But it does *look* like that it would be appropriate to give precedence to tls1_suiteb even in SSL_OP_PRIORITIZE_CHACHA case.

File: ssl/s3_lib.c (hunk 8 above)
Comment (dot-asm) [reply ssl/s3_lib.c]: Obviously
```
int num = sk_SSL_CIPHER_num(x)
for (i = 0; i < num; i++) {}
```

**Comment by dot-asm** [comment]
Note that it needs re-base.

File: ssl/s3_lib.c (hunk 6 above)
Comment (tmshort) [reply ssl/s3_lib.c]: re: tls1_suiteb; I don't care, just give me a definitive answer. :)

Comment (dot-asm) [reply ssl/s3_lib.c]: Among other things RFC6460 says that "A Suite B TLS server ... MUST accept ... the TLS_ECDHE_ECDSA_WITH_AES_* cipher suite *if* it is offered in the ClientHello message." Well, there still is certain ambiguity, because when it comes to client side it says that "One of these two cipher suites MUST be the first (most preferred) cipher suites in the ClientHello message." So that Chacha being first in list presented in ClientHello formally disqualifies it as compliant, and then one can wonder if server is in turn obliged to comply with the requirement... I'd say in case of doubt assume the worst, i.e. make server comply. Or in other words don't SSL_OP_PRIORITIZE_CHACHA in tls1_suiteb case.

File: ssl/s3_lib.c (hunk 7 above)
Comment (tmshort) [reply ssl/s3_lib.c]: Done as a separate commit

File: ssl/s3_lib.c (hunk 6 above)
Comment (tmshort) [reply ssl/s3_lib.c]: Done

File: ssl/s3_lib.c [hunk 9]
```diff
@@ -4145,9 +4148,52 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+                if (found) {
+                    new_chacha = sk_SSL_CIPHER_new_null();
+                    /* if reserve fails, then there's likely a memory issue */
+                    if (new_chacha != NULL && sk_SSL_CIPHER_reserve(new_chacha, num)) {
+                        /* Put ChaCha at the top */
+                        for (i = 0; i < num; i++) {
+                            c = sk_SSL_CIPHER_value(srvr, i);
+                            if (c->algorithm_enc == SSL_CHACHA20POLY1305)
+                                sk_SSL_CIPHER_push(new_chacha, c);
+                        }
```
Comment (dot-asm) [inline ssl/s3_lib.c]: ??? You already have `c` to push, you don't need this loop at all, just sk_SSL_CIPHER_push(new_chacha, c)...

File: ssl/s3_lib.c [hunk 10]
```diff
@@ -4120,6 +4120,9 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
     STACK_OF(SSL_CIPHER) *prio, *allow;
     int i, ii, ok;
     unsigned long alg_k = 0, alg_a = 0, mask_k = 0, mask_a = 0;
+#ifndef OPENSSL_NO_CHACHA
+    STACK_OF(SSL_CIPHER) *new_chacha = NULL;
```
Comment (dot-asm) [inline ssl/s3_lib.c]: Wouldn't you find `prio_chacha` being a better name for this variable?

File: ssl/s3_lib.c (hunk 9 above)
Comment (tmshort) [reply ssl/s3_lib.c]: Yes, it's needed, there can be more than one ChaCha20 cipher in the server list, and they all need to be put at the start. The first loop at 4168~4174 attempts to avoid allocating the new stack, if possible.

File: ssl/s3_lib.c (hunk 10 above)
Comment (tmshort) [reply ssl/s3_lib.c]: sure

File: ssl/s3_lib.c (hunk 9 above)
Comment (tmshort) [reply ssl/s3_lib.c]: I suppose I could shorten the loop by pushing the found cipher first; then iterating through the rest.

**Comment by dot-asm** [comment]
Does the fact that it's two commits mean that this in fact is nominated for 1.1.0? I *for one* would vote against it in present state, because it's new *feature* [specifically visible from source code]. Note "in present state" as I *for one* would consider even implicit prioritization acceptable, i.e. with no enable flag. But that would be surely considered too radical by others :-)

**Comment by richsalz** [comment]
I think it's hard to see how this could be a bugfix, and therefore it's master-only.

File: ssl/s3_lib.c (hunk 9 above)
Comment (dot-asm) [reply ssl/s3_lib.c]: Ah! You mean that there are other fields in client's head of list and goal is to prioritize *all* Chachas. One could omit this loop if the match was exact...

Comment (tmshort) [reply ssl/s3_lib.c]: We're still doing server priority; there's a decent chance (>=50% given the number of ChaCha20 ciphers) that they won't match, but another ChaCha20 cipher in the list will.

Comment (dot-asm) [reply ssl/s3_lib.c]: Question is if prioritizing *all* Chachas is actually what client wants. Would maybe starting to push ciphers from client list till it's not Chacha, and then append server list omitting all Chachas more adequate?

Comment (dot-asm) [reply ssl/s3_lib.c]: No, this would be problematic, right? Because in this case you wouldn't be able to interchange prio and allow...

Comment (tmshort) [reply ssl/s3_lib.c]: We still want server priority, not client priority; so we want to only include those ChaCha ciphers in the server list, in the order that the server has them.

Comment (dot-asm) [reply ssl/s3_lib.c]: All right, rename new_chacha to prio_chacha, *optionally* start loop from i+1 instead of 0 (you'll have to push c first), and add comment explaining that we pull up all Chacha ciphers.

**Comment by tmshort** [comment]
@InfoHunter suggested that it might be nice to add to 1.1.0; a comment on the use of sk_reserve indicated that it should be done separately() so as to allow a 1.1.0 backport. And there's PR #3922, which would also be a backport of a feature.

File: ssl/s3_lib.c (hunk 9 above)
Comment (tmshort) [reply ssl/s3_lib.c]: Done

**Comment by tmshort** [comment]
ping @openssl/committers?

**Comment by snhenson** [comment]
Might we want to prioritise other mobile friendly ciphersuites in future? If so then SSL_OP_PRIORITIZE_CHACHA_FOR_MOBILE might be better named SSL_OP_PRIORITIZE_FOR_MOBILE

**Comment by dot-asm** [comment]
FOR_MOBILE was actually removed, so it's just PRIORITIZE_CHACHA, and it's perfectly suitable name, because that's what it checks for.

**Comment by tmshort** [comment]
Rebased to fix conflict. Still looking for second approval @openssl/committers.

**Comment by InfoHunter** [comment]
So it's decided not to include this into 1.1.0?

**Comment by tmshort** [comment]
@InfoHunter no idea, but it's set up so it could be.
@openssl/committers, still waiting on reviews

**Comment by InfoHunter** [comment]
> @openssl/committers, still waiting on reviews

Seems '@openssl/committers' doesn't work, could be: @openssl

**Comment by mattcaswell** [comment]
@openssl/committers  should work?

**Comment by tmshort** [comment]
The @openssl/committers shows up bolded on your post, but not mine? Likely because it's a private group, so you can see it, but the rest of us can't.

**Comment by InfoHunter** [comment]
> The @openssl/committers shows up bolded on your post, but not mine? Likely because it's a private group, so you can see it, but the rest of us can't.

Yes. And that is also not prompted when typing in the text area of Github. I asked @richsalz before and we figured out the `@openssl` stuff and it worked :-)

**Comment by tmshort** [comment]
Still waiting on second approval @openssl

**Comment by InfoHunter** [comment]
I agree with @tmshort . This should be more efficiently handled...

**Comment by tmshort** [comment]
Still waiting on second review... @openssl

File: ssl/s3_lib.c [hunk 11]
```diff
@@ -4145,9 +4148,53 @@ const SSL_CIPHER *ssl3_choose_cipher(SSL *s, STACK_OF(SSL_CIPHER) *clnt,
...
+                int found = 0;
+                for (i = 0; i < num; i++) {
+                    c = sk_SSL_CIPHER_value(srvr, i);
+                    if (c->algorithm_enc == SSL_CHACHA20POLY1305) {
+                        found = 1;
+                        break;
+                    }
+                }
+                if (found) {
+                    prio_chacha = sk_SSL_CIPHER_new_null();
```
Comment (paulidale) [inline ssl/s3_lib.c]: `sk_SSL_CIPHER_new_reserve(NULL, num)` would be the preferred allocation & reservation method now.

**Comment by paulidale** [comment]
Still needs an OMC review.

**Comment by tmshort** [comment]
@paulidale doesn't @dot-asm count as OMC?

File: ssl/s3_lib.c (hunk 11 above)
Comment (tmshort) [reply ssl/s3_lib.c]: I assume you aren't asking for a change, as the sk_reserve() API is used below.

Comment (InfoHunter) [reply ssl/s3_lib.c]: I guess so, since the `new_reserve` API was added after this pull request.

File: apps/apps.h [hunk 12]
```diff
@@ -233,6 +234,8 @@ int set_cert_times(X509 *x, const char *startdate, const char *enddate,
             "Disallow initial connection to servers that don't support RI"}, \
         {"allow_no_dhe_kex", OPT_S_ALLOW_NO_DHE_KEX, '-', \
             "In TLSv1.3 allow non-(ec)dhe based key exchange on resumption"}, \
+        {"prioritize_chacha", OPT_S_PRIORITIZE_CHACHA, '-', \
+            "Prioritize ChaCha ciphers for mobile clients"}, \
```
Comment (vdukhovni) [inline apps/apps.h]: The wording here should also change to drop "mobile" and just say "when preferred by client" or some such

Comment (vdukhovni) [reply apps/apps.h]: Also perhaps shorten "prioritize" to "prefer"

File: doc/man1/s_server.pod [hunk 13]
```diff
@@ -510,6 +511,10 @@ Disable RFC4507bis session ticket support.
 
 Use the server's cipher preferences, rather than the client's preferences.
 
+=item B<-prioritize_chacha>
+
+Prioritize ChaCha ciphers for mobile clients.
```
Comment (vdukhovni) [inline doc/man1/s_server.pod]: Ditto re mobile

File: doc/man3/SSL_CONF_cmd.pod [hunk 14]
```diff
@@ -171,6 +171,12 @@ Use server and not client preference order when determining which cipher suite,
 signature algorithm or elliptic curve to use for an incoming connection.
 Equivalent to B<SSL_OP_CIPHER_SERVER_PREFERENCE>. Only used by servers.
 
+=item B<-prioritize_chacha>
+
+Prioritize ChaCha ciphers when the client has a ChaCha20 cipher at the top of
+its preference list. This usually indicates a mobile client is in use.
```
Comment (vdukhovni) [inline doc/man3/SSL_CONF_cmd.pod]: Mention that this option has no effect unless server preference is enabled. Replace "mobile" with client likely has no hardware AES support.  Perhaps mention that server performance may be impacted.

File: doc/man3/SSL_CONF_cmd.pod [hunk 15]
```diff
@@ -382,21 +388,26 @@ B<Bugs>: enable various bug workarounds. Same as B<SSL_OP_ALL>.
...
-B<ServerPreference> use server and not client preference order when
+B<ServerPreference>: use server and not client preference order when
 determining which cipher suite, signature algorithm or elliptic curve
 to use for an incoming connection.  Equivalent to
 B<SSL_OP_CIPHER_SERVER_PREFERENCE>. Only used by servers.
 
-B<NoResumptionOnRenegotiation> set
+B<PrioritizeChaCha>: prioritizes ChaCha ciphers when the client has a
+ChaCha20 cipher at the top of its preference list. This usually indicates
+a mobile client is in use. Equivalent to B<SSL_OP_PRIORITIZE_CHACHA>.
```
Comment (vdukhovni) [inline doc/man3/SSL_CONF_cmd.pod]: Ditto re mobile.

File: apps/apps.h (hunk 12 above)
Comment (tmshort) [reply apps/apps.h]: "Prefer ChaCha ciphers when preferred by client" sounds awkward to me...

File: doc/man3/SSL_CTX_set_options.pod [hunk 16]
```diff
@@ -180,6 +180,14 @@ messages, and ignore renegotiation requests via ClientHello.
 In TLSv1.3 allow a non-(ec)dhe based key exchange mode on resumption. This means
 that there will be no forward secrecy for the resumed session.
 
+=item SSL_OP_PRIORITIZE_CHACHA
+
+When SSL_OP_CIPHER_SERVER_PREFERENCE is set, temporarily reprioritize
+ChaCha20-Poly1305 ciphers to the top of the server cipher list if a
+ChaCha20-Poly1305 cipher is at the top of the client server list. This helps
+mobile clients use ChaCha20-Poly1305 if that cipher is anywhere in the sever
```
Comment (vdukhovni) [inline doc/man3/SSL_CTX_set_options.pod]: s/sever/server/
Replace mobile with more general language.

File: doc/man3/SSL_CONF_cmd.pod (hunk 14 above)
Comment (tmshort) [reply doc/man3/SSL_CONF_cmd.pod]: The second statement implies x86_64 architecture, which is not always the case. I don't see any documentation referencing hardware AES support, and no mention of relative performance of cipher suites. i.e. the documentation doesn't specifically state anywhere that choosing ChaCha20 over AES in SSL_set_cipher() will impact server performance. But there's also a counter argument, that ChaCha may provide a boost in performance as far as the client is concerned. I don't think it's the place in the API documentation to go into over-arching performance issues that are better described in a whitepaper or something.

**Comment by vdukhovni** [comment]
I am not convinced that this is a good long-term interface. It feels somewhat like a quick hack, and we have a few too many of these as a legacy. Perhaps implementing cipher equivalence classes makes more sense, that way the server can order lists of classes that ignore the distinction between AESGCM and CHACHA, or even between AES128 and AES256, and between SHA2-256 and SHA2-384 and chooses the first class in which there is a common cipher.  Within that class we can then go with the client preference or the server preference.  That is the server preference bit becomes two bits whose class to choose, and whose cipher to choose within that class.

**Comment by richsalz** [comment]
I think adding equivalence classes and further complicating that syntax is also not optimal.  This is a relatively clean, although not generic, solution to a big existing problem:  Mobile wants chacha and desktop wants AES.

**Comment by vdukhovni** [comment]
> @richsalz I think adding equivalence classes and further complicating that syntax is also not optimal. This is a relatively clean, although not generic, solution to a big existing problem: Mobile wants chacha and desktop wants AES.
The syntax need not be made more complicated, this could be an opportunity to simplify it.  We can add an "@V=1" to the front of the cipher string to indicate a brand new syntax that better reflects 20 years of experience. The "not generic" part is precisely the problem here.  This is a stop-gap, not an architectural feature. And yet we may be stuck with it for a long time once introduced. We should I think at least discuss alternative designs.

**Comment by richsalz** [comment]
I understand the issue; nobody would talk about it before, when the equivalence classes issue was raised multiple times.  I'd rather see site-wide cipher profiles (from RedHat) as a better solution.  But nobody is stepping forward to do that -- in fact, attempts to address this, in many ways, have been ignored or rebuffed so far.

Tnis code should go into the next release.

**Comment by vdukhovni** [comment]
Often doing nothing is better than doing something that lacks a solid design.  If this issue needs to be addressed, let's prioritize it and put on the roadmap, and see some design proposals  before code is written.  This is a major part of the public API, one that many users find confusing, and this is probably as good a time as any to address it. We've managed without a ChaCha optimization for some time, it can wait if need be.

**Comment by richsalz** [comment]
Disagree, and we'll leave it at that.

**Comment by t-j-h** [comment]
There is strong demand for this capability ... and pushing it off simply means we end up with a whole range of approaches that get put into production to achieve the same effect - leading to slower updates and more custom patches for something that is relatively straight forward and focused. It isn't generalised - but we also haven't seen strong demand for a generalised solution. 

@vdukhovni do you feel strongly enough to -1 this? 
I'm ready to approve it - but if you'd prefer to take it to an OMC vote I'm fine with that too.

**Comment by tmshort** [comment]
Tweaked the documentation.
Fixed to use sk_SSL_CIPHER_new_reserve()

**Comment by vdukhovni** [comment]
> @t-j-h @vdukhovni do you feel strongly enough to -1 this? I'm ready to approve it - but if you'd prefer to take it to an OMC vote I'm fine with that too.

I feel strongly that whoever does feel the itch strongly enough to actually address an issue is on the hook for doing it right and not saddling the project with a legacy of short-term work-arounds.  So once we have a pull request it should high-quality design, not just high-quality implementation.

In this case, I don't think an up/down on vote on whether to merge the change would really address my concerns. I'd like to suggest we take a short pause break to consider a more ambitious design that makes it easier to optimize the cipher order using a syntax that most mortals can understand and one that avoids out-of-band control bits (like prioritize chacha) and puts all the controls in the cipher string.

**Comment by vdukhovni** [comment]
Perhaps the author of this PR is willing to propose a more radical overhaul after considering how this is solved in BoringSSL, GnuTLS, TLS toolkits in other languages, ... and proposing something that is realistic and combines some of the best features of the existing approaches.

**Comment by tmshort** [comment]
Given the nature of the request; it is doubtful that it could be completed in any reasonable about of time; and would likely miss the 1.1.1 timeframe. The existing cipher specification is arcane and easily broken; especially where forward and backward compatibility are required.
This option is not incompatible with any possible proposed preference list mechanism that might be implemented in the future.

TL;DR: no.

**Comment by t-j-h** [comment]
If no one steps forward to tackle a more comprehensive solution I'll add approval to this tomorrow so @paulidale or @dot-asm can merge.

**Comment by vdukhovni** [comment]
> @tmshort Given the nature of the request; it is doubtful that it could be completed in any reasonable about of time; and would likely miss the 1.1.1 timeframe. The existing cipher specification is arcane and easily broken; especially where forward and backward compatibility are required. This option is not incompatible with any possible proposed preference list mechanism that might be implemented in the future.

This is disappointing, we should not be in such a hurry that we introduce long-term API extensions that are layered on top of weak foundations, rather than revisit the design.  Sure someone might do something better "some day", but "some day" rarely comes.  The time to ask for better care is when the patient is opened up and the surgeon has the scalpel in hand.

The current design with internal stacks manipulated by a series of push/pop operations that move around groups of ciphers while maintaining the internal order of the groups, and putting removed ciphers back in the order in which they removed is much too arcane. Translating a preference for (kECDHE, kEDH, kRSA) with (aRSA, aECDSA) with (AESGCM, AES, ...) ordered by cipher strength into a cipherlist requires fancy gymnastics.

It should instead be possible to state that all other things being equal one prefers some key exchange algorithms over (or the same as) others, some public key algorithms over (or the same as) others, some bulk ciphers over others, ... and be able to set up equivalences which might go with the client's preference when the server otherwise does not care.  It is time to create a new cipherlist string syntax, and the new to address new use-cases is a good time to take that on.  Surely we can do better than what we've got, and do so in fact in a reasonable amount of time.

**Comment by tmshort** [comment]
@t-j-h ?

**Comment by paulidale** [comment]
Merged to master.
